src/Module/BaseApi.php

```diff
@@ -195,80 +187,108 @@ public static function getRequest(array $defaults)
 	 *               'authenticated' => return status,
 	 *               'user_record' => return authenticated user record
 	 */
-	protected static function login(string $scope)
+	protected static function login(string $scope): bool
 	{
-		if (empty(self::$current_user_id)) {
-			self::$current_token = self::getTokenByBearer();
-			if (!empty(self::$current_token['uid'])) {
-				self::$current_user_id = self::$current_token['uid'];
-			} else {
-				self::$current_user_id = 0;
-			}
+		if (!self::getCachedCurrentUserIDFromRequest()) {
+			Logger::debug(API_LOG_PREFIX . 'failed', ['module' => 'api', 'action' => 'login', 'parameters' => $_SERVER]);
+			header('WWW-Authenticate: Basic realm="Friendica"');
+			throw new HTTPException\UnauthorizedException('This API requires login');
 		}
 
-		if (!empty($scope) && !empty(self::$current_token)) {
-			if (empty(self::$current_token[$scope])) {
-				Logger::warning('The requested scope is not allowed', ['scope' => $scope, 'application' => self::$current_token]);
-				DI::mstdnError()->Forbidden();
-			}
-		}
-
-		if (empty(self::$current_user_id)) {
-			// The execution stops here if no one is logged in
-			api_login(DI::app());
+		$current_token = self::getCachedTokenByBearer();
+		if ($current_token && empty($current_token[$scope])) {
+			Logger::warning('The requested scope is not allowed', ['scope' => $scope, 'application' => self::getCachedTokenByBearer()]);
+			DI::mstdnError()->Forbidden();
 		}
 
-		self::$current_user_id = api_user();
-
-		return (bool)self::$current_user_id;
+		return (bool)api_user();
 	}
 
 	/**
-	 * Get current application
+	 * Returns the current logged user id from the local variable or attempts to retrieve it from the request.
 	 *
-	 * @return array token
+	 * @return int
+	 * @throws HTTPException\ForbiddenException
+	 * @throws HTTPException\InternalServerErrorException
+	 * @throws HTTPException\UnauthorizedException
 	 */
-	protected static function getCurrentApplication()
+	public static function getCachedCurrentUserIdFromRequest(): int
 	{
-		return self::$current_token;
+		static $current_user_id = null;
+
+		if (!is_null($current_user_id)) {
+			return $current_user_id;
+		}
+
+		$current_user_id = self::getCurrentUserIDFromRequest();
+
+		return $current_user_id;
 	}
 
 	/**
 	 * Get current user id, returns 0 if not logged in
 	 *
 	 * @return int User ID
+	 * @throws HTTPException\ForbiddenException
+	 * @throws HTTPException\InternalServerErrorException
+	 * @throws HTTPException\UnauthorizedException
 	 */
-	public static function getCurrentUserID(bool $nologin = false)
+	private static function getCurrentUserIDFromRequest(): int
```

Comment:
I don't understand the reason for renaming this function. It just fetches the current user id from various sources, so why not keep it short and simple?

Reply:
Because there are at least 3 methods retrieving the current user ID from different sources. They all depend on each other, granted, but they all are different:
- `getCachedCurrentUserIdFromRequest` depends on `getCurrentUserIdFromRequest` but only once, hence the `Cached` part.
- `getCurrentUserIdFromRequest` depends on `getCurrentUserIDFromToken` but can also perform `api_login` if it wasn't sufficient.
- `getCurrentUserIDFromToken` only retrieves the current User ID from the Bearer token.

Reply:
For me `FromRequest` in that function name doesn't tell anything. To make it clearer it would had to be called `getCurrentUserIdFromTokenOrBasicAuthDependingOnTheLoginStyle()`. I think that this information can be put in the function documentation instead.

Also since the non caching part is a private function and thus cannot be called from outside the class, there is no choice using either the one or the other function. So the information whether this is cached or not doesn't add additional benefit. It's not like in `ParseUrl` where we have the two functions `getSiteinfoCached` and `getSiteinfo` that can be used to either fetch the cached data or fresh data.

Reply:
Ideally, the non-caching method should be used for unit testing, and so it probably shouldn't be private, but I haven't written tests for it yet.

In my view `FromRequest` means that is doesn't come from a local session for example, it is using the data coming from the general HTTP request instead of any local source.

Reply:
Technically a request with a provided token and a request with a session is really similar. In both cases the HTTP header contains some key value. In the one case the "Authorization" header, in the other case it's the "Session" header. And in both cases we have to locally fetch that provided id value. That means that there is not really that huge difference.

Reply:
It’s true, but there is a difference that would be evident with dependency injection: the two method wouldn’t be passed the same dependencies. Since we are still relying on superglobals, I believe we must be even more explicit about this kind of difference in the name of the methods since there aren’t any parameters to distinguish them.

In most other object-oriented programming languages, we could define `getCurrentUserId(Request $request)` and `getCurrentUserId(Session $session)` but in PHP we can’t, so we have to be explicit in the method name itself.

Reply:
To be honest: This sounds for me like another argument against this "dependency injection" stuff.

Reply:
Given that it tries to make everything as explicit as possible at the cost of some verbosity, I can see why you would think that. 😉